rfd/0212-jsonpath-interpolation.md

```diff
@@ -0,0 +1,584 @@
+feature is necessary for Teleport to integrate with these OIDC solutions.
+
+## Details
+
+### `jsonpath` Function
+
+The `jsonpath` function will be added as another standard trait expression
+function, making it usable in `login_rule` trait maps/expressions, role label
+expressions (e.g. `node_labels_expression`), etc.
+
+However, traits cannot have arbitrary JSON object values. The only exception
+will be during login when OIDC claims are mapped to traits and login rules are
+applied.
+
+Note: currently, login rules are applied after the claims to traits mapping.
+Instead, we will apply the login rules to the OIDC claims (`map[string]any`)
+before mapping them to traits (`map[string][]string`).
+
+Note: the `jsonpath` function can *technically* still be used for role label
+expressions, since strings and string lists can still be treated as json. It's
+unlikely for there to be a real use case for this, though some of the complex
+features, such as filtering with numerical comparisons, could be used to extend
+the label expression functionality currently available.
+
+#### JSONPath Libraries
+
+While the JSONPath query language has an [official RFC](https://www.rfc-editor.org/rfc/rfc9535.html)
+as of 2024, the inception of the query language was in an [article](https://goessner.net/articles/JsonPath/)
+written in 2007. The original article left many questions unanswered, causing
+the JSONPath language to morph in several directions as different JSONPath
+projects came up with their own answers.
+
+While the RFC has helped to realign these disparate projects, there is much
+more work to be done, as can be clearly seen by this [comparison project](https://cburgmer.github.io/json-path-comparison/).
+
+According to the comparison project, https://github.com/ohler55/ojg is
+currently the closest to the RFC specification out of the Go projects listed
+there.
+
+By using a project close to the specification, we can selectively rely on
+official JSONPath documentation, like those linked above, as well as the
+[online sandboxes](https://serdejsonpath.live/).
+
+Note: We must be careful when updating the upstream library, as some of the
+syntax may be subject to change as it adjusts to the RFC and some of the
+unresolved syntax disagreements in the community.
+
+#### JSONPath Syntax
+
+Before continuing to read this RFD, you should familiarize yourself with the
+basics of [JSONPath syntax](https://support.smartbear.com/alertsite/docs/monitors/api/endpoint/jsonpath.html).
+
+When going through the JSONPath examples, you may find it useful to run the
+queries in a [sandbox](https://serdejsonpath.live/).
+
+### OIDC Claims
+
+During OIDC login, a user's OIDC claims serve two purposes:
+
+* Determining what Teleport roles to give the user, using the `claims_to_roles` mapping.
+* Setting claims as Teleport user traits.
+  * Login rules are applied to these traits before they are set for the user.
```

Comment:
might be worth noting here that login rules are applied before the claims_to_roles mapping and their output can affect that mapping